src/mesh/mesh_optimizer.hh

```diff
@@ -0,0 +1,437 @@
+        const Vec3 dimensions = arma::vec3(bounding_box.max() - bounding_box.min());
+        normalizer_ = Normalizer(bounding_box.min(), std::max({dimensions[0], dimensions[1], dimensions[2]}));
+    }
+
+    inline void calculate_node_curve_values_as_hilbert() {
+        node_refs_.reserve(mesh_->n_nodes());
+        for (uint i = 0; i < mesh_->n_nodes(); ++i) {
+            node_refs_.emplace_back(i, hilbert_value(normalizer_.normalize(mesh_->nodes_.vec<3>(i)), normalizer_.normalize(node_sizes_[i]), mesh_->element_vec_[i].dim()));
+        }
+    }
+
+    inline void calculate_node_curve_values_as_zcurve() {
+        node_refs_.reserve(mesh_->n_nodes());
+        for (uint i = 0; i < mesh_->n_nodes(); ++i) {
+            node_refs_.emplace_back(i, zcurve_value(normalizer_.normalize(mesh_->nodes_.vec<3>(i)), normalizer_.normalize(node_sizes_[i]), mesh_->element_vec_[i].dim()));
+        }
+    }
+
+    inline void calculate_node_curve_values_as_mean_of_coords() {
+        node_refs_.reserve(mesh_->n_nodes());
+        for (uint i = 0; i < mesh_->n_nodes(); ++i) {
+            const Vec3 tmp_norm = normalizer_.normalize(mesh_->nodes_.vec<3>(i));
+            node_refs_.emplace_back(i, (tmp_norm[0] + tmp_norm[1] + tmp_norm[2]) / 3);
+        }
+    }
+
+    inline void calculate_node_curve_values_as_first_coord() {
+        node_refs_.reserve(mesh_->n_nodes());
+        for (uint i = 0; i < mesh_->n_nodes(); ++i) {
+            const Vec3 tmp_norm = normalizer_.normalize(mesh_->nodes_.vec<3>(i));
+            node_refs_.emplace_back(i, tmp_norm[0]);
+        }
+    }
+
+    inline void calculate_node_curve_values_as_obtained_from_elements() {
+        node_refs_.reserve(mesh_->n_nodes());
+        for (uint i = 0; i < mesh_->n_elements(); ++i) {
+            const Element& el = mesh_->element_vec_[element_refs_[i].original_index_];
+            for (uint j = 0; j < el.dim() + 1; ++j) {
+                node_refs_.emplace_back(el.nodes_[j], element_refs_[i].curve_value_);
+            }
+        }
+    }
+
+    inline void calculate_element_curve_values_as_mean_of_nodes() {
+        element_refs_.reserve(mesh_->n_elements());
+        for (uint i = 0; i < mesh_->n_elements(); ++i) {
+            const std::array<uint, 4>& nodeIndexes = mesh_->element_vec_[i].nodes_;
+            double tmp_sum = 0;
+            for (uint j = 0; j < mesh_->element_vec_[i].dim() + 1; ++j) {
+                tmp_sum += node_refs_[nodeIndexes[j]].curve_value_;
+            }
+            element_refs_.emplace_back(i, tmp_sum / (mesh_->element_vec_[i].dim() + 1));
+        }
+    }
+
+    inline void calculate_element_curve_values_as_hilbert_of_centers() {
+        element_refs_.reserve(mesh_->n_elements());
+        for (uint i = 0; i < mesh_->n_elements(); ++i) {
+            element_refs_.emplace_back(i, hilbert_value(normalizer_.normalize(ElementAccessor<3>(mesh_, i).centre()), normalizer_.normalize(element_sizes_[i]), mesh_->element_vec_[i].dim()));
```

Comment:
Use existing element measures in hilbert and zcurve calculations, remove own calculation of element sizes.

Reply:
Is it method ElementAccessor::measure ?